### PR TITLE
HDDS-11723. Tool to better micro benchmark hbase performance in Ozone

### PR DESCRIPTION
**Pull request by duongkame**
## What changes were proposed in this pull request?

See https://issues.apache.org/jira/browse/HDDS-11723.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-11723

## How was this patch tested?

CI.

### CODE REVIEW

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HsyncGenerator.java [hunk 1]
```diff
@@ -16,35 +16,39 @@
...
  * Example usage:
  *
- * To generate 1000 hsync calls with 10 threads on a single file:
- *    ozone freon hsync-generator -t 10 --bytes-per-write=1024 -n 1000
- *
- * To generate 1000 hsync calls with 10 threads on 3 files simultaneously:
+ * To simulate hlog that generates 1M hsync calls with 5 threads:
  *
- *    ozone freon hsync-generator -t 10 --bytes-per-write=1024 --number-of-files=3 -n 1000
+ *    ozone freon hlog-emulator -t 5 --writes-per-transaction=32 --bytes-per-write=8 -n 1000000
```
Comment (smengcl) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HsyncGenerator.java:51]: looks like the subcommand name hasn't changed

```suggestion
 *    ozone freon hsync-generator -t 5 --writes-per-transaction=32 --bytes-per-write=8 -n 1000000
```

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HsyncGenerator.java [hunk 2]
```diff
@@ -87,55 +97,75 @@ public HsyncGenerator() {
...
     timer = getMetrics().timer("hsync-generator");
     data = PayloadUtils.generatePayload(writeSize);
 
+    startTransactionWriter();
+
     try {
       runTests(this::sendHsync);
     } finally {
-      for (FSDataOutputStream outputStream : outputStreams) {
         outputStream.close();
```
Comment (smengcl) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HsyncGenerator.java:122]: nit

```suggestion
      outputStream.close();
```

**Review by smengcl — APPROVED** [review]
Thanks @duongkame . lgtm

@jojochuang Do you also want to take a quick look?

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HsyncGenerator.java [hunk 3]
```diff
@@ -87,55 +97,75 @@ public HsyncGenerator() {
...
-      outputStream.write(data);
-      outputStream.hsync();
-      callsPerFile[i].incrementAndGet();
-      return null;
+      while (true) {
+        int transaction = writtenTransactions.take();
+        int lastSyncedTransaction = this.lastSyncedTransaction.get();
+        if (transaction > lastSyncedTransaction) {
+          outputStream.hsync();
+          this.lastSyncedTransaction.compareAndSet(lastSyncedTransaction, transaction);
```
Comment (adoroszlai) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HsyncGenerator.java:165]:
```
hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HsyncGenerator.java
 162: 'lastSyncedTransaction' hides a field.
```

```suggestion
        int lastSynced = lastSyncedTransaction.get();
        if (transaction > lastSynced) {
          outputStream.hsync();
          lastSyncedTransaction.compareAndSet(lastSynced, transaction);
```

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HsyncGenerator.java (hunk 2 above)
Comment (adoroszlai) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HsyncGenerator.java:122]:
```
hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HsyncGenerator.java
 122: 'finally' child has incorrect indentation level 8, expected level should be 6.
```

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HsyncGenerator.java [hunk 4]
```diff
@@ -53,32 +57,38 @@
...
   @Option(names = {"--bytes-per-write"},
       description = "Size of each write",
-      defaultValue = "1024")
+      defaultValue = "8")
   private int writeSize;
 
-  @Option(names = {"--number-of-files"},
-      description = "Number of files to run test.",
-      defaultValue = "1")
-  private int numberOfFiles;
```
Comment (adoroszlai) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HsyncGenerator.java:70]: Test still uses this option:

https://github.com/apache/ozone/blob/3340947e8d2e5a96f88e8be320e64d4c46991674/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/freon/TestHsyncGenerator.java#L103

and fails:

```
AssertionFailedError: expected: <0> but was: <2>
  ...
	at org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:531)
	at org.apache.hadoop.ozone.freon.TestHsyncGenerator.test(TestHsyncGenerator.java:106)
```

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HsyncGenerator.java [hunk 5]
```diff
@@ -53,32 +57,38 @@
...
+  @Option(names = {"--writes-per-transaction"},
+      description = "Size of each write",
+      defaultValue = "32")
+  private int writesPerTransaction;
```
Comment (chungen0126) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/HsyncGenerator.java:79]: Could we also add this option to the command in the test to make sure it works correctly?

**Review by chungen0126 — COMMENTED** [review]
Thanks @duongkame for the patch.

**Comment by duongkame** [comment]
Thanks for reviewing @smengcl @adoroszlai @chungen0126.

**Comment by smengcl** [comment]
Merged. Thanks @duongkame for the patch. Thanks @adoroszlai @chungen0126 for reviewing this as well.

**Pull request merged**